pgxn/neon/unstable_extensions.c

```diff
@@ -0,0 +1,183 @@
+#include <assert.h>
+#include <stdlib.h>
+#include <string.h>
+
+#include "postgres.h"
+
+#include "nodes/plannodes.h"
+#include "nodes/parsenodes.h"
+#include "tcop/utility.h"
+#include "utils/errcodes.h"
+#include "utils/guc.h"
+
+#include "unstable_extensions.h"
+
+#if PG_MAJORVERSION_NUM < 16
+#define guc_free(ptr) free(ptr)
+#define guc_malloc(level, size) malloc(size)
+#define guc_strdup(level, s) strdup(s)
+#endif
+
+typedef struct UnstableExtensions
+{
+	size_t	extc;
+	char   *extv[FLEXIBLE_ARRAY_MEMBER];
+} UnstableExtensions;
+
+static bool					allow_unstable_extensions = false;
+static char				   *unstable_extensions_str = NULL;
+static UnstableExtensions  *unstable_extensions = NULL;
+
+static ProcessUtility_hook_type PreviousProcessUtilityHook = NULL;
+
+static bool
+check_unstable_extensions(char **newval, void **extra, GucSource source)
+{
+	char			   *n;
+	UnstableExtensions *exts;
+	char			   *curr_ext;
+	size_t				max_extc;
+
+	if (*newval == NULL || (*newval)[0] == '\0')
+		return true;
+
+	/* At this point, we know we have at least 1 extension like: ext */
+	max_extc = 1;
+	for (size_t i = 0; i < strlen(*newval); ++i)
+	{
+		if ((*newval)[i] == ',')
+			++max_extc;
+	}
+
+	exts = guc_malloc(ERROR, sizeof(*exts) + max_extc * sizeof(char *));
+
+	n = guc_strdup(ERROR, *newval);
+	while ((curr_ext = strsep(&n, ",")))
+	{
+		/* In the event, we receive a config like: ",,ext" */
+		if (curr_ext[0] == '\0')
+			continue;
+
+		exts->extv[exts->extc++] = guc_strdup(ERROR, curr_ext);
+	}
+
+	guc_free(n);
+
+	*extra = exts;
+
+	return true;
+}
+
+static void
+assign_unstable_extensions(const char *newval, void *extra)
+{
+	unstable_extensions = extra;
+}
+
+static void
+CheckUnstableExtension(
+	PlannedStmt *pstmt,
+	const char *queryString,
+	bool readOnlyTree,
+	ProcessUtilityContext context,
+	ParamListInfo params,
+	QueryEnvironment *queryEnv,
+	DestReceiver *dest,
+	QueryCompletion *qc)
+{
+	const char *ext_name = NULL;
+	Node	   *parseTree = pstmt->utilityStmt;
+
+	if (allow_unstable_extensions || unstable_extensions == NULL)
+		goto process;
+
+	switch (nodeTag(parseTree))
+	{
+		case T_AlterExtensionStmt:
+		{
+			AlterExtensionStmt *stmt = castNode(AlterExtensionStmt, parseTree);
+
+			ext_name = stmt->extname;
+			break;
+		}
+		case T_AlterExtensionContentsStmt:
+		{
+			AlterExtensionContentsStmt *stmt = castNode(AlterExtensionContentsStmt, parseTree);
```

Comment:
More of a question: I'm not sure we need to do anything with alter. If you already installed it, you kinda gave a consent that it's unstable